1주차 과제 초기 셋팅-GET 조회 및 POST 시 Task추가 구현

euijeonghwang opened the pull request:
URI가 /tasks포함 시 TaskHttpHandler를 타도록 구현함.
requestMethod가 GET일 경우 Task 모델 전체 조회하도록 구현 및 POST일 경우 Task List에 추가하도록 구현함.

euijeonghwang committed:
1주차 과제 초기 셋팅-GET 조회 및 POST 시 Task추가 구현

URI가 /tasks포함 시 TaskHttpHandler를 타도록 구현함.
requestMethod가 GET일 경우 Task 모델 전체 조회하도록 구현 및 POST일 경우 Task List에 추가하도록 구현함.

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,21 @@
     public static void main(String[] args) {
+
+        try {
+            HttpServer httpServer = HttpServer.create(new InetSocketAddress(8000), 0);
```
8000은 아마 포트 번호를 나타내고 있는 것 같고, 0이라는 숫자는 backlog값으로 시스템 기본값을 쓰는 것 같아요.
다른 개발자들이 숫자만 봐서는 이 숫자의 의미가 무엇인지 이해하기 어렵겠어요. 이 값들이 단순히 숫자가 아니라 특별한 의미를 갖는다는 것을 상수로 표현해보는 것은 어떨까요?

See also
  - https://refactoring.com/catalog/replaceMagicLiteral.html

hannut91 commented on app/src/test/java/com/codesoom/assignment/model/TaskTest.java:
```diff
@@ -0,0 +1,13 @@
+class TaskTest {
+
+
+}
```
코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.

![image](https://user-images.githubusercontent.com/14071105/200711698-cdb5542a-e913-48c2-8900-b7bfcbf5fc08.png)


이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님의 블로그 글을 참고해 주세요

See also

  - [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0)

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,34 @@
+
+    public void setTitle(String title) {
+        this.title = title;
+        this.id = ++Task.num;
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Task {
+
+    private static int num;
```
이 `num`은 어디에 사용되는 값인가요?

euijeonghwang replied:
`Task`에 하나 추가될 때마다 이 `num`을 사용해서 id값을 1씩 증가하고자 했던 의도였습니다만, 위에 적어주신 코멘트와 더불어서 조금 더 고민해보겠습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/model/TaskTest.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
코드숨 과정을 진행하는 동안에는 import *의 사용을 지양하도록 합시다. 가급적 어떤 것을 import 하는지 명시해 줄 수 있도록 해봅시다.

intellij 설정에서 자동으로 `import *`으로 변경되지 않도록 하는 방법도 있는데 찾아서 한 번 적용해 보는 것도 좋겠습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+     * @throws IOException
+     */
+    private String tasksToJson() throws IOException {
+        OutputStream outputStream = new ByteArrayOutputStream();
+        objectMapper.writeValue(outputStream, tasks);
+        return outputStream.toString();
```
변수가 변한다는 사실 자체만으로도 프로그램의 복잡성을 많이 증가시킵니다. 어떤 값을 내가 사용할 때 이 값이 현재 어떤 값인지 예측하기 어렵기 때문이죠. 따라서 `final`을 사용해서 재할당 할 수 없도록 하면 좋을 것 같습니다.

물론 여기서는 좀 단순해서 필요없어 보이지만, 왜 final을 사용해야 하는지도 한 번 공부해 봅시다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+     * @throws JsonProcessingException
+     */
+    private Task contentToTask(String content) throws JsonProcessingException {
+        Task newTask = objectMapper.readValue(content, Task.class);
+        return newTask;
```
좌변에 변수가 코드를 이해하는데 도움을 주지 못하는 것 같습니다. `contentToTask`의 반환 타입에 이미 `Task`가 있으므로 바로 반환해도 좋을 것 같습니다.

```suggestion
        return objectMapper.readValue(content, Task.class);
```
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+
+        String content = "";
+
+        if (requestMethod.equals("GET")) {
```
`GET, POST` 같은 문자열을 enum으로 사용해 보는 것은 어떨까요?

euijeonghwang replied:
반영하여 enum 으로 만들었습니다. 그런데 requestMethod와 비교하려면 toString()이나 name() 메서드를 사용해야 하는데 오히려 더 복잡하게 구현되는건 아닌지 궁금합니다!

hannut91 replied:
그럴 수 있겠네요 ㅎㅎ 그렇다면 다음 영상을 통해서 개선하는 과정을 한 번 보고 따라해 봅시다

https://www.youtube.com/watch?v=qRZL2NkXRdo

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        if (requestMethod.equals("POST") && !body.isBlank()) {
+            System.out.println("requestMethod.equals(\"POST\") && !body.isBlank()");
```
아마 위의 결과를 확인해 보려고 출력하셨던 것 같습니다. 확인이 끝났다면, 코드에서 삭제하는게 좋을 것 같습니다

euijeonghwang replied:
혹시 slf4j 라이브러리를 추가해서 사용해도 될까요? 과제 시 다른 라이브러리 사용 가능여부도 궁금합니다.

hannut91 replied:
로그를 통해서 확인하고 싶으셨군요. 다른 외부 라이브러리를 사용해도 괜찮습니다. 다만 과제를 제출하기 위해서 한다기 보다 다른 팀원들과 협업을 하는 상황이라고 가정하고 라이브러리를 도입해 보세요. 그럴려면 이 라이브러리를 쓰는 이유를 충분히 다른 팀원들에게 설명해 주어야 할거예요

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+
+        String content = "";
+
+        if (requestMethod.equals("GET")) {
```
이 if문에 들어온다면, 아래 if문에는 절대로 들어갈 일이 없을 것 같습니다. 그렇다면 이 if문에서 하는 일이 끝나면, 다른 코드들은 읽을 필요 없이 할려면 다른 방법을 사용해야 될 것 같습니다. 

See also
  - https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,86 @@
+        String content = "";
+
+        if (requestMethod.equals("GET")) {
+            content = tasksToJson();
```
이 함수는 굉장히 추상화가 잘 되어있습니다. 잘은 모르겠지만, tasks를 json으로 바꿔주겠지~ 라고 이해할 수 있습니다. 하지만 위의 다른 코드들은 추상화 레벨이 굉장히 낮습니다. 각 데이터를 어떻게 처리할지 전부 기술되어 있죠.

이렇게 추상화 레벨이 다르면 코드를 이해하기 어렵게 만듭니다. 따라서 위의 코드들도 추상화하여 이해하기 쉬운 코드를 만들어 보세요.

hannut91 reviewed (commented):
잘 부탁드립니다!

euijeonghwang committed:
HTTP 메서드를 enum으로 사용해라
GET, POST, DELETE, PUT, PATCH http메서드를 enum 클래스로 만듦

euijeonghwang committed:
HttpMethod별로 조건문 구성 및 공통 변수는 상수로 만들어라

if문으로 GET, POST, DELETE, PUT/PATCH 구분함
코드 리뷰 반영:
 - App.java : 포트 번호 및 backlog값 상수로 표현, 포트 8000 -> 8080 변경
 - TaskHttpHandler.java : OutputStream final로 변경. contentToTask() 커밋받은거 받고 반영하려고 했으나 git을 잘몰라서 그냥 수동으로 수정함.
 - 코드 마지막 빈 줄 추가하도록 intellij 설정
 - import single class로 변경

euijeonghwang committed:
조건에 맞게 필터링된 Task를 가져오는 테스트를 만들어라

Stream filter()를 활용하여 조건에 맞는 필터링을 하였는지 테스트함.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,126 @@
+    private final OutputStream outputStream = new ByteArrayOutputStream();
+    private final List<Task> tasks = new ArrayList<>();
+
+    private static final Integer SC_OK = 200;
```
`SC`는 아마 Status Code를 나타내는 것 같아요. 아주 자명한 것이 아니라면 줄임말은 사용하지 않는게 좋을 것 같아요

euijeonghwang replied:
HttpServletResponse 인터페이스에 나와있는 변수명을 사용했었습니다. 보내주신 링크의 영상 보고 객체화 시키겠습니다 👍🏻

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,126 @@
+     * @throws IOException
+     */
+    private String getFilteredTask(String taskId) throws IOException {
+        //TODO stream filter 찾아보기
+        objectMapper.writeValue(outputStream, tasks.stream()
+                                                    .filter(t -> Integer.parseInt(taskId) == t.getId()));
```
stream을 사용했는데, 정확히 알아보고 싶으셔서 TODO를 적어놓으셨군요 ㅎㅎ 공부한 것을 채널에 공유해 주세요~!

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,42 @@
+        this.id = ++Task.num;
+    }
+
+    @Override
+    public String toString() {
+        return "Task{" +
+                "id=" + id +
+                ", title='" + title + '\'' +
+                '}';
+    }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```kotlin
return "{ id = $id, title = $title }"
```

euijeonghwang replied:
찾아보니 String format이 "+"로 연결하는 것보다 느리지만, format을 사용하는 가장 큰 이유는 localized하기 편하기 때문인 것 같습니다. 이렇게 이해하면 될까요?

hannut91 replied:
그런 이유도 있었군요 ㅎㅎ 저는 단순히 더 보기 좋아서 말씀드려봤어요

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,126 @@
+        // DELETE /tasks/{id}
+        if (requestMethod.equals(DELETE.name())) {
+            if (!resource.isBlank()) {
+                tasks.remove(Integer.parseInt(resource)-1);
```
여기서 -1 이 의미하는 것이 무엇일까요?

euijeonghwang replied:
Task의 id는 1부터 시작하는데, List의 remove는 0부터 시작하는 index를 파라미터로 요구해서 -1을 해서 맞추고자 했던 의도였습니다,,

euijeonghwang replied:
생각해보니 이 로직은 `IndexOutOfBoundsException` 이 발생할 가능성이 있네요!!

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,126 @@
+        String requestMethod = exchange.getRequestMethod(); //GET, POST, PUT/PATCH, DELETE...
+        String uri = exchange.getRequestURI().getPath();
+        //TODO 어떻게 하면 /posts/{id} 에서 {id}부분만 가져올 수 있는지 찾기
+        String[] path = uri.split("/");
```
path를 `/`로 나누었군요. 나누어진 결과가 path라고 하는 것 같은데, 이걸 정확히 무엇이라고 불리는지 찾아보고 정확한 단어를 쓰는 것이 좋을 것 같아요

euijeonghwang committed:
Task의 toString()을 String.format()을 활용한 문자열을 리턴하라

Task 클래스의 toString() 메서드에서 + 연산자로 연결한 것을 String클래스의 format() 메서드를 활용하여 리턴하는 것으로 변경함

euijeonghwang committed:
HTTP 메서드에 따른 handle 메서드를 따로 빼라

id값 여부에 따른 로직 분리
Get 전체조회, Post 완료
Delete는 삭제는 잘 되나 커밋 작성하면서 remove시 문제가 있음을 인지함
 - tasks에 id는 3인 task가 하나 있는데  -1을 하면 index의 값이 2가 나오므로 에러 발생!!

euijeonghwang committed:
removeTask 테스트해라

IndexOutofBoundsException 나는 것을 확인함. 로직 수정 필요

euijeonghwang committed:
static대신 Response를 자바 객체로 변환해라

Response abstract class를 만들고, Created(201), NotFound(404), Success(200) 자식클래스 생성함

euijeonghwang committed:
HTTP 메서드에 따른 과제 구현

App.java : 테스트 포트 8000으로 변경
Task.java : Task의 id Long으로 변경, id값은 외부에서 1씩 증가하도록 수정
TaskHttpHandler.java : 넘어오는 id값에 따라 GET/PUT/PATCH/DELETE 구현

euijeonghwang committed:
id값이 오지 않을 때 404에러를 발생시켜라

Delete시 id값이 오지 않는다면 404에러 리턴

euijeonghwang commented:
테스트 관련하여 질문있습니다,,
테스트가 다 실패하고 아래와 같은 에러가 발생하는데 제가 짠 코드의 문제인지 아니면 환경설정문제인지 아무리 찾아봐도 모르겠습니다.
jdk11버전 사용하던 걸 15로 맞춰봤는데도 안되네요,,
os는 mack Ventura 13.0입니다

`tasks › GET /tasks/:id › with existing task id › responses task

    FetchError: request to http://localhost:8000/tasks failed, reason: socket hang up

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)

  ● tasks › GET /tasks/:id › with existing task id › responses task

    FetchError: request to http://localhost:8000/tasks/undefined failed, reason: read ECONNRESET

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)

  ● tasks › GET /tasks/:id › with not existing task id › responses 404 Not Found error

    FetchError: request to http://localhost:8000/tasks/1 failed, reason: socket hang up

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)

  ● tasks › POST /tasks › responses 201 Created

    FetchError: request to http://localhost:8000/tasks failed, reason: socket hang up

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)

  ● tasks › PUT /tasks/:id › with existing task id › responses updated task

    FetchError: request to http://localhost:8000/tasks failed, reason: socket hang up

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)`

hannut91 commented:
제가 접근했던 방식은 이렇습니다.

일단 테스트를 실행해보았더니 GET과 POST는 성공하는데, PUT, PATCH 그리고 DELETE만 실패하고 있었습니다.

그래서 일단 API호출은 잘 되나 수동으로 테스트 해봤는데, 잘 되는 것을 확인했습니다. 그런데 테스트 결과에서는 404라고 응답하고 있습니다.

테스트를 확인해보니 

```js
  describe('PATCH /tasks/:id', () => {
    let id;

    context('with existing task id', () => {
      beforeEach(async () => {
        const res = await frisby.post('/tasks', { title });
        const task = JSON.parse(res.body);
        id = task.id; // <=== 응답결과로 생성된 id를 응답하기를 기대하고 있습니다.
      });

      it('responses updated task', async () => {
        // 생성된 id로 수정을 요청하고 있습니다.
        const res = await frisby.put(`/tasks/${id}`, { title: '밥 먹기' })
          .expect('status', 200);
        const task = JSON.parse(res.body);

        expect(task.title).toBe('밥 먹기');
      });
    });
});
```

task 하나를 생성한 후 생성된 id로 수정하는 테스트로 작성되어 있는데, 생성할 때 생성된 id를 응답하는 것이 아니라 할 일 목록을 응답하고 있어서 id가 undefined였고 그래서 404 응답이 오는 것이었습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,198 @@
+     * @return 1씩 증가된 id
+     */
+    private Long generatedId() {
+        return id += 1;
```
return 되는 값이 증가되기 전의 값인가요 아니면 증가된 후의 값인가요? 저라면 명확히 쓸 것 같습니다.
```suggestion
        id += 1;
        return id;
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,198 @@
+     * @return 필터링된 Task
+     * @throws IOException
+     */
+    private Task getFilteredTask(Long taskId) {
+        //TODO stream filter 찾아보기
+       return tasks.stream()
+               .filter(t -> t.getId().equals(taskId))
+               .findFirst()
+               .orElse(null);
+    }
```
`tasks`에서 우리가 원하는 task를 찾지 못할 경우 `null`을 반환하고 있는 것 같아요.이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

euijeonghwang replied:
제가 Optional을 들어만 봤지, 이렇게 사용하는 적은 처음이라 어떻게 활용해야되는지 찾아보고 있습니다,, 그렇다면 `orElseThrow()`를 사용한다면 `NPE`를 던져서 

```
if (lastSegment.startsWith("0") || getFilteredTask(Long.parseLong(lastSegment)) == null) {
    handleError(exchange, content);
    return;
}
```
이 부분에서 `try-catch` 구문으로 예외를 잡거나 또는 `Supplier`를 활용하여 404 에러를 반환하는 Exception을 구현한 `orElseThrow()`를 주어야 할까요?
아니면 `getFilteredtask` 에서 `handleError()`를 호출하기 위해서 `HttpExchange`를 파라미터로 넘겨주어야할까요?
어떤 방식이 더 실무에서 많이 활용하는 방법인지 질문드립니다!

hannut91 replied:
`orElseThrow`를 사용한다면, 여기서 발생하는 예외를 던져주고, 그 예외를 처리하는 무언가가 있어야겠죠? 이건 차차 다음 강의에서 배우실 수 있으실거예요 ㅎㅎ

여기서는 `orElse(null)`을 하지 않고 아예 `Optional`타입 그대로 반환할 수도 있습니다. 그러면 이 함수를 사용하는 입장에서는 Optional을 처리하지 않으면, 값을 사용할 수 없기 때문에, 빈 값 확인을 `강제`하게 됩니다. 하지만 `null`은 그렇지 않죠. 사용하는 입장에서 null을 처리해야 한다는 것을 알 수 없습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,198 @@
+     * @return true || false
+     */
+    private boolean isNumeric(String value) {
+        return value != null && value.matches("[0-9.]+");
```
정규식을 이용해서 숫자인지 판별하고 있는 것 같아요. 만약에 `12xxx` 이렇게 들어온 값은 어떻게 될까요?

euijeonghwang replied:
그 경우에는 `false`를 리턴하게됩니다. 그러나 실제로 실행했을 경우 전체 목록을 `GET`해오기 때문에 이 부분에 대한 로직이 필요한 것 같습니다!

euijeonghwang committed:
코드리뷰 generateID() 반영

명확한 id 리턴하도록 수정

euijeonghwang commented:
> 제가 접근했던 방식은 이렇습니다.
> 
> 일단 테스트를 실행해보았더니 GET과 POST는 성공하는데, PUT, PATCH 그리고 DELETE만 실패하고 있었습니다.
> 
> 그래서 일단 API호출은 잘 되나 수동으로 테스트 해봤는데, 잘 되는 것을 확인했습니다. 그런데 테스트 결과에서는 404라고 응답하고 있습니다.
> 
> 테스트를 확인해보니
> 
> ```js
>   describe('PATCH /tasks/:id', () => {
>     let id;
> 
>     context('with existing task id', () => {
>       beforeEach(async () => {
>         const res = await frisby.post('/tasks', { title });
>         const task = JSON.parse(res.body);
>         id = task.id; // <=== 응답결과로 생성된 id를 응답하기를 기대하고 있습니다.
>       });
> 
>       it('responses updated task', async () => {
>         // 생성된 id로 수정을 요청하고 있습니다.
>         const res = await frisby.put(`/tasks/${id}`, { title: '밥 먹기' })
>           .expect('status', 200);
>         const task = JSON.parse(res.body);
> 
>         expect(task.title).toBe('밥 먹기');
>       });
>     });
> });
> ```
> 
> task 하나를 생성한 후 생성된 id로 수정하는 테스트로 작성되어 있는데, 생성할 때 생성된 id를 응답하는 것이 아니라 할 일 목록을 응답하고 있어서 id가 undefined였고 그래서 404 응답이 오는 것이었습니다

잘 이해가 안갑니다..😿 저도 console로 찍어서 확인해보니 id가 `undefined`로 뜨네요..
혹시 형태가 `{..}`가 아니라 `[{...}]`로 들어와서 못찾아오는걸까요? 직접 수동테스트할 땐 id가 있으면 `{...}`형태로 잘 들어오는데 모르겠습니다.

그리고 GET이랑 POST는 테스트가 잘된다고 하셨는데 저는 테스트 실행하면 11건 전체 다 실패한다고 나옵니다. 
GET이랑 POST가 잘 된다면 저한테도 몇개는 성공으로 떠야하는거 아닌가요,?
- node : v19.0.1
- npm: 8.19.2
입니다
localhost사용대신 127.0.0.1을 사용해보라고 해도 동일합니다..

euijeonghwang committed:
isNumeric() 조건 추가 및 handleUpdate() 메서드 대신 handleGetDetail() 사용

0으로 시작하거나 해당 id에 해당하는 찾고자 하는 Task가 없을 때 error발생시키기
ex) 01, 023, 12xxx 등..

PUT/PATCH 시 handleUpdate 메서드 대신 handleGetDetail()하도록 수정함

hannut91 commented:
`POST /tasks`로 할 일을 만들었을 때, 응답하는 결과의 모양이 다음과 같습니다.

```
{
  "id": 1
}
```

그런데 위와 같이 응답이 오지 않아서 테스트가 실패하고 있습니다. 할 일을 생성할 떄 올바른 응답을 주도록 수정해야 합니다
